### PR TITLE
canvas test code

### PR DESCRIPTION
**Pull request by gdenes355**
from Paul:
added graphics pane, stub for canvas messages and example c04

### CODE REVIEW

File: src/challenge/Challenge.tsx [hunk 1]
```diff
@@ -56,16 +58,24 @@ class Challenge extends React.Component<ChallengeProps, ChallengeState> {
...
   printCallback = throttle(
     () => this.setState({ consoleText: this.currentConsoleText }),
     100
   );
 
+  drawCallback = throttle(
+    () => this.setState({ graphicsCommand: this.currentGraphicsCommand }),
+    100
+  );  
+
```
Comment (gdenes355) [inline src/challenge/Challenge.tsx:72]: I don't think this should be communicating the command via setState (which triggers a React redraw). Using throttle will prevent performance issues, but you might lose some calls (I think that's what is happening with the arc drawing).

Consider just setting up a reference to the Graphics pane (see how it's done for PyEditor), and call a method on it.

File: src/challenge/Challenge.tsx [hunk 2]
```diff
@@ -92,6 +103,11 @@ class Challenge extends React.Component<ChallengeProps, ChallengeState> {
     this.printCallback();
   }
 
+  draw(text: string) {
+    this.currentGraphicsCommand = text;
+    this.drawCallback();
+  }
+
```
Comment (gdenes355) [inline src/challenge/Challenge.tsx:110]: once the throttle is gone, this can be probably deleted

File: src/challenge/Challenge.tsx [hunk 3]
```diff
@@ -56,16 +58,24 @@ class Challenge extends React.Component<ChallengeProps, ChallengeState> {
...
+    () => this.setState({ graphicsCommand: this.currentGraphicsCommand }),
+    100
+  );  
+
   state: ChallengeState = {
     starterCode: null,
     savedCode: null,
     worker: null,
     consoleText: "Press debug to get started...",
+    graphicsCommand: "fillRect(0,0,0,0)",
```
Comment (gdenes355) [inline src/challenge/Challenge.tsx:78]: once the throttle is gone, this can be deleted.

File: src/challenge/Challenge.tsx [hunk 4]
```diff
@@ -356,11 +386,20 @@ class Challenge extends React.Component<ChallengeProps, ChallengeState> {
...
                 </Allotment.Pane>
+                <Allotment.Pane
+                  visible={this.getVisibilityWithHack(
+                    !this.state.editorFullScreen
+                  )}
+                  maxSize={400}
+                  minSize={400}
+                >
+                  {this.renderGraphicsPane()}
+                </Allotment.Pane>
```
Comment (gdenes355) [inline src/challenge/Challenge.tsx:402]: OK for proof of concept, but I wonder if maybe we should have a challenge typ(e) which is `graphics`, andonly show this for that type.
Also, could consider showing graphics in place of the console or tabbed.

File: src/components/GraphicsPane.tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+import { Box } from "@mui/material";
+import "./GraphicsPane.css";
+import Context from "@mui/lab/node_modules/@mui/base/TabsUnstyled/TabsContext";
+
+type GraphicsPaneProps = {
+  content: string;
+  onInterrupt: () => void;
+};
+
+const GraphicsPane = (props: GraphicsPaneProps) => {
```
Comment (gdenes355) [inline src/components/GraphicsPane.tsx:11]: To get rid of the state-based communication, you might need to convert this into a React class (see PyEditor, ParsonsEditor, Challenge). 
React functional components are great due to their simplicity, but for your case a class (with a runCommand method) makes more sense.

File: public/static/js/init.py [hunk 6]
```diff
@@ -28,21 +142,55 @@ def clear(self):
...
 test_output = TestOutput()
 
 active_breakpoints = set()
 test_inputs = []
 test_outputs = []
 step_into = False
 
+# setup turtle library
+
+with open("turtle.py", "w") as file:
```
Comment (gdenes355) [inline public/static/js/init.py:156]: a bit confused by this

**Review by gdenes355 — COMMENTED** [review]
If we can please shorten / remove test code from src/components/CanvasDisplay.tsx, then I think this is now good to go and can be merged

File: src/components/CanvasDisplay.tsx [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+};
+
+type CanvasDisplayState = {
+  turtle_x_init: number,
+  turtle_y_init: number,    
+  turtle_x: number,
+  turtle_y: number,
+  turtle_dir: number
+  turtle: any
+};
```
[the remark this answers is not in the document]
Comment (bakerpdgit) [reply src/components/CanvasDisplay.tsx]: I've removed these now.

File: public/static/js/init.py (hunk 6 above)
Comment (bakerpdgit) [reply public/static/js/init.py:156]: There may be a better way but to try to get PLS / python compatibility, by creating a turtle.py on the pyodide temporary session file system, it allows you to "import turtle" as part of your python example code and then create a turtle 
t = turtle.turtle().

A similar suggestion in the teams conversation regarding the tk stub to allow key press messages to be passed back to bouncing ball in a python compatible fashion.

**Comment by bakerpdgit** [comment]
I'm good with merging this back into main if you are happy to.

**Pull request merged**